### PR TITLE
[Inductor] Fix CPU bool argmin/argmax incorrect indices

### LINKED ISSUE #184893: [Inductor] argmin/min returns incorrect indices for boolean tensors
### 🐛 Describe the bug

This seems to be similar to #174069 fixed in #174076 so cc @ezyang @gchanan @kadeng @msaroufim @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @aditew01 @chauhang @penguinwu @voznesenskym @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @muchulee8 @amjames @aakhundov @coconutruben @jataylo @Aminsed @jansel 
But it affects (also?) CPU.

I made a quick reproducer iterating over sizes:

```python
import torch
device = 'cpu'

def fnmax(x):
    return x.max(0)
def fnmin(x):
    return x.min(0)

for n in range(1, 8):
    t1 = torch.randint(2, size=(n, 32), dtype=torch.bool, device=device)
    for desc, fn in [('min', fnmin)]:
        print(f"Testing {desc} {n=}...")
        eager = fn(t1)
        compiled = torch.compile(fn, backend='inductor')(t1)
        eager_values = [t1[j][i].item() for i, j in enumerate(eager.indices)]
        compiled_values = [t1[j][i].item() for i, j in enumerate(compiled.indices)]
        print("\tEager values:   ", eager_values)
        print("\tCompiled values:", compiled_values)
        check = "PASS" if eager.indices.tolist() == compiled.indices.tolist() else "FAIL"
        print(f"\t{check}")
```

<details><summary>Output</summary>
<p>
```
Testing min n=1...
	Eager values:    [False, False, False, False, True, True, False, True, False, True, False, True, False, False, True, False, True, False, True, False, True, False, True, True, False, False, True, False, False, True, True, True]
	Compiled values: [False, False, False, False, True, True, False, True, False, True, False, True, False, False, True, False, True, False, True, False, True, False, True, True, False, False, True, False, False, True, True, True]
	PASS
Testing min n=2...
	Eager values:    [True, True, False, False, False, True, False, False, False, True, False, False, True, True, False, True, False, False, False, True, True, False, False, False, False, False, False, False, False, True, False, False]
	Compiled values: [True, True, True, False, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True]
	FAIL
Testing min n=3...
	Eager values:    [False, False, False, False, False, True, False, True, True, False, False, False, False, False, False, False, False, False, True, False, False, False, False, False, False, False, False, False, False, False, True, False]
	Compiled values: [True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, False, True, True, True, False, True, True, False]
	FAIL
Testing min n=4...
	Eager values:    [False, False, False, False, False, False, False, False, False, False, False, False, True, False, False, True, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False]
	Compiled values: [True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True]
	FAIL
Testing min n=5...
	Eager values:    [False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False]
	Compiled values: [True, True, True, True, True, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True]
	FAIL
Testing min n=6...
	Eager values:    [False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False]
	Compiled values: [True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, True, True]
	FAIL
Testing min n=7...
	Eager values:    [False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, True, False, False, False, False, False, False, False, False]
	Compiled values: [True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True]
	FAIL
```
</p>
</details> 

What looks odd: It seems to start behaving like `max` instead of `min`.

It also works (i.e. fails) with a modified example from #174069 and random inputs:

```python
import torch
import torch._dynamo

device = 'cpu'

def fn(x):
    return x.min(1, keepdim=True)[1]

mask = torch.randint(2, size=(4, 8, 32), dtype=torch.bool)

result_eager = fn(mask)
torch._dynamo.reset()
result_compiled = torch.compile(fn, backend='inductor')(mask)

e_item = result_eager[0, 0, 0].item()
c_item = result_compiled[0, 0, 0].item()
print(e_item, c_item, e_item == c_item)  # Outputs e.g. "1 0 False"
print(result_compiled)
print(result_eager)
print(f'Eager result: {result_eager[0, 0, 0].item()}')
print(f'Inductor result: {result_compiled[0, 0, 0].item()}')
```

I found that as the newly added test `python test/inductor/test_torchinductor.py CpuTests.test_max_min_bool_cpu` fails for me

### Versions

PyTorch version: 2.12.0+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Rocky Linux 9.6 (Blue Onyx) (x86_64)
GCC version: (GCC) 14.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.34

Python version: 3.13.5 (main, Jul 30 2025, 17:50:55) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-5.14.0-570.49.1.el9_6.x86_64-x86_64-with-glibc2.34
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: 
GPU models and configuration: 
GPU 0: NVIDIA H100
GPU 1: NVIDIA H100
GPU 2: NVIDIA H100
GPU 3: NVIDIA H100

Nvidia driver version: 580.159.04
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           52 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  64
On-line CPU(s) list:                     0-63
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 9334 32-Core Processor
CPU family:                              25
Model:                                   17
Thread(s) per core:                      1
Core(s) per socket:                      32
Socket(s):                               2
Stepping:                                1
Frequency boost:                         enabled
CPU(s) scaling MHz:                      96%
CPU max MHz:                             3910.2529
CPU min MHz:                             1500.0000
BogoMIPS:                                5400.52
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good amd_lbr_v2 nopl xtopology nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba perfmon_v2 ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif x2avic v_spec_ctrl vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid overflow_recov succor smca fsrm flush_l1d debug_swap
L1d cache:                               2 MiB (64 instances)
L1i cache:                               2 MiB (64 instances)
L2 cache:                                64 MiB (64 instances)
L3 cache:                                256 MiB (8 instances)
NUMA node(s):                            8
NUMA node0 CPU(s):                       0-7
NUMA node1 CPU(s):                       8-15
NUMA node2 CPU(s):                       16-23
NUMA node3 CPU(s):                       24-31
NUMA node4 CPU(s):                       32-39
NUMA node5 CPU(s):                       40-47
NUMA node6 CPU(s):                       48-55
NUMA node7 CPU(s):                       56-63
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Mitigation; Safe RET
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsx async abort:           Not affected

Versions of relevant libraries:
[pip3] numpy==2.3.2
[pip3] nvidia-cublas==13.1.0.3
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-nvrtc-cu12==12.6.77
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.5.1.17
[pip3] nvidia-cudnn-cu13==9.20.0.48
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.6.3
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-nccl-cu12==2.26.2
[pip3] nvidia-nccl-cu13==2.29.7
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx==13.0.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] torch==2.12.0
[pip3] triton==3.7.0
[conda] Could not collect

### PR DESCRIPTION
**Pull request by psiwho**
Fixes #184893

Summary:

When using `torch.compile` with the Inductor backend on CPU, `max(dim)[1]`
(argmax) and `min(dim)[1]` (argmin) could return incorrect indices for
boolean tensors.

Root cause:

CPU C++ codegen represents vectorized boolean values as `VecMask<float, N>`.
In that representation, true lanes are all-bits-set mask payloads rather than
numeric `1` values. For bool argmin/argmax reductions, this can send the
comparison through the floating arg-reduction path, where true lanes can be
treated like NaN-ish float mask values instead of ordinary boolean values.

Patch:

This extends the bool argmin/argmax fix from #174076 to CPU C++ codegen,
so boolean arg reductions compare `int32` values internally (`False -> 0`,
`True -> 1`) before selecting indices. Public result dtypes are unchanged.

Tests:

The existing `test_max_min_bool` test now runs on CPU as well as CUDA, using
deterministic CPU repro shapes for #184893. The stale CPU dynamic-shapes skip
for `test_max_min_bool_dynamic_shapes` is also removed.

Test plan:

- `python test/inductor/test_torchinductor.py CpuTests.test_max_min_bool_cpu -v`
- `python test/inductor/test_torchinductor_dynamic_shapes.py DynamicShapesCpuTests.test_max_min_bool_dynamic_shapes_cpu -v`

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by Flamefire** [comment]
Question as you mentioned vectorization: Wouldn't it be better to cast to uint8_t for a processing 4 times the values per loop or a specialization handling bools as bitmasks and using vector intrinsics to gather set/unset bits? Or even compiler intrinsics to get first set/unset bit and loop over those.

**Comment by psiwho** [comment]
@Flamefire Good points! I looked into this further after you mentioned it. Even casting to uint8_t requires adjusting tiling to fully saturate the SIMD lanes, which is more involved than I initially thought. The bitmask / intrinsic path is more powerful but a bigger lift still.

I'm inclined to keep this PR focused and tackle the vectorization separately, but happy to hear if you think it's worth pursuing here.

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Review by jansel — CHANGES_REQUESTED** [review]
fix test failures

**Review by jansel — APPROVED** [review]

**Comment by jansel** [comment]
@claude explain those CI failures, are they related to this PR?

**Review by jansel — CHANGES_REQUESTED** [review]
fix test failures

**Comment by Flamefire** [comment]
Conflicts due to https://github.com/pytorch/pytorch/pull/185970 being merged addressing the same issue. Maybe still check if anything from here is not included yet
